feat(jdbc): add index to commonly queried fields in the WHERE conditions of the triggers table

gluttonweb opened the pull request:
<!-- Thanks for submitting a Pull Request to Kestra. To help us review your contribution, please follow the guidelines below:

- Make sure that your commits follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification e.g. `feat(ui): add a new navigation menu item` or `fix(core): fix a bug in the core model` or `docs: update the README.md`. This will help us automatically generate the changelog.
- The title should briefly summarize the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share a flow example to help the reviewer understand and QA the change.
- Use "closes" to automatically close an issue. For example, `closes #1234` will close issue #1234. -->

### What changes are being made and why?

Changes: 
1. Add an index to the next_execution_date field in the triggers table
2. Add performance time metrics

Reason: 
* The next_execution_date field, which is frequently used in WHERE conditions of common SQL queries on the triggers table, does not have an index. Under high load conditions, an index needs to be added to improve query performance and the performance of the core handler() method

<!-- Please include a brief summary of the changes included in this PR e.g. closes #1234. -->

---

### How the changes have been QAed?
![image](https://github.com/user-attachments/assets/2f8c424c-c522-4d62-b14e-dcca5bad853e)
<!-- Include example code that shows how this PR has been QAed. The code should present a complete yet easily reproducible flow.

```yaml
# Your example flow code here
```

Note that this is not a replacement for unit tests but rather a way to demonstrate how the changes work in a real-life scenario, as the end-user would experience them.

Remove this section if this change applies to all flows or to the documentation only. -->

Commit:
feat: add index to commonly queried fields in the WHERE conditions of the triggers table

loicmathieu reviewed (changes requested):
Hi,
Thanks a lot for your contribution!
You added a metric for the database query, but we already have metrics for all queries thanks to a jOOQ interceptor so it didn't seems to bring added value.
However, I notice that we didn't have duration metrics for the `handle` method itself, it may be good to add it.

Commit:
feat: add duration metrics for handle() method

gluttonweb commented:
> Hi, Thanks a lot for your contribution! You added a metric for the database query, but we already have metrics for all queries thanks to a jOOQ interceptor so it didn't seems to bring added value. However, I notice that we didn't have duration metrics for the `handle` method itself, it may be good to add it.

Okay, thank you very much for your feedback. I have removed the current SQL query duration metric and added duration metrics for the handle() method (we actually had this in our test environment, but it wasn't included in this PR previously).

loicmathieu commented on core/src/main/java/io/kestra/core/schedulers/AbstractScheduler.java:
```diff
@@ -517,7 +518,7 @@ private void handle() {
             return;
         }
-
+        Instant startInstant = Instant.now();
         ZonedDateTime now = now();
```
I think you can use the ZonedDateTime `now` instead of creating an instant.

loicmathieu commented on core/src/main/java/io/kestra/core/schedulers/AbstractScheduler.java:
```diff
@@ -660,6 +661,9 @@ private void handle() {
                 });
         });
+        metricRegistry
+            .timer(MetricRegistry.METRIC_SCHEDULER_HANDLE_DURATION, MetricRegistry.METRIC_SCHEDULER_HANDLE_DURATION_DESCRIPTION, Lists.newArrayList().toArray(new String[0]))
```
```suggestion
            .timer(MetricRegistry.METRIC_SCHEDULER_HANDLE_DURATION, MetricRegistry.METRIC_SCHEDULER_HANDLE_DURATION_DESCRIPTION)
```

loicmathieu reviewed (commented):
Can you also add the index to PostgreSQL and H2?
Please, keep the two commits as I plan to backport the missing indices to 0.22 if possible.

Commit:
feat: synchronize next_execution_date field index in triggers table to PostgreSQL and H2

loicmathieu reviewed (approved):
LGTM, thanks.

Commit:
fix: fix bug in H2 index creation statement

gluttonweb commented:
> LGTM, thanks.

gluttonweb commented:
> LGTM, thanks.

I just fixed a bug in the H2 index creation. Everything looks good now~

gluttonweb commented:
> > LGTM, thanks.
> 
> I just fixed a bug in the H2 index creation. everything looks good now~

loicmathieu commented:
Great! Thanks!

loicmathieu commented:
@gluttonweb H2 seems still to cause issue, can you have a look?

Commit:
fix: Duration metrics for the handle() method: both start and end times use ZonedDateTime

gluttonweb commented:
> @gluttonweb H2 seems still to cause issue, can you have a look?

It appears that the handle() method duration metrics were using ZonedDateTime for the start time but Instant for the end time. The error occurred because Instant doesn't contain timezone information, causing the calculation to fail. I've submitted a new fix commit.

loicmathieu reviewed (approved):
LGTM

loicmathieu commented:
@gluttonweb I merged too fast your PR and had to revert and re-create it as we now have strong commit policy ;) 
I recreated https://github.com/kestra-io/kestra/pull/8398 with only two commits, you will still be author of the commit.

gluttonweb commented:
> @gluttonweb I merged too fast your PR and had to revert and re-create it as we now have strong commit policy ;) I recreated #8398 with only two commits, you will still be author of the commit.

Thank you very much! I'm extremely excited as this is my first PR contribution to the Kestra open source project. We will continue to contribute in the future and look forward to more intellectual exchanges.

loicmathieu commented:
@gluttonweb great new! We love new contributors ;) 
Don't hesitate to ask any question in our slack, we have a contributors channel dedicated to contribution discussion.
